ODIN_II/SRC/netlist_create_from_ast.cpp

```diff
@@ -3469,51 +3472,166 @@ signal_list_t* create_operation_node(ast_node_t* op, signal_list_t** input_lists
     oassert(input_port_width != -1);
     oassert(output_port_width != -1);
 
-    for (i = 0; i < list_size; i++) {
-        if ((operation_node->type == SR) || (operation_node->type == ASL) || (operation_node->type == SL) || (operation_node->type == ASR)) {
-            /* Need to check that 2nd operand is constant */
-            ast_node_t* second = op->children[1];
-            if (second->type != NUMBERS)
-                error_message(NETLIST, op->loc, "%s", "Odin only supports constant shifts at present\n");
-            oassert(second->type == NUMBERS);
-
-            /* for shift left or right, it's actually a one port operation. The 2nd port is constant */
-            if (i == 0) {
+    /* allocate the pins for the ouput port */
+    allocate_more_output_pins(operation_node, output_port_width);
+    add_output_port_information(operation_node, output_port_width);
+
+    if ((operation_node->type == SR) || (operation_node->type == ASL) || (operation_node->type == SL) || (operation_node->type == ASR)) {
+
+        for (int j = 0; j < input_lists[0]->count; j++) {
+            add_pin_to_signal_list(shift_input_pins[0], input_lists[0]->pins[j]);
+            input_lists[0]->pins[j] = NULL;
+        }
+        /* allocate the pins needed */
+        allocate_more_input_pins(operation_node, input_port_width);
+        /* record this port size */
+        add_input_port_information(operation_node, input_port_width);
+        /* hookup the input pins */
+        /* 
+        *   if the input is signed, then instead of adding pad 0 for extension we have to add its sign bit as extension bits
+        */
+        hookup_input_pins_from_signal_list(operation_node, current_idx, shift_input_pins[0], 0, input_port_width, verilog_netlist);
+        
+        if (op->children[1]->type != NUMBERS) {
+
+            current_idx += input_port_width;
+
+            for (int j = 0; j < input_lists[1]->count; j++) {
+                add_pin_to_signal_list(shift_input_pins[1], input_lists[1]->pins[j]);
+                input_lists[1]->pins[j] = NULL;
+            }
+            /* allocate the pins needed */
+            allocate_more_input_pins(operation_node, log2(op->children[1]->types.vnumber->get_value()));
```

Comment:
op->children[1]->types.vnumber->get_value() ? why are you getting the value?  line 3495 says its not a number